server/mcp/user_client.go

```diff
@@ -0,0 +1,284 @@
+	serverID string
+}
+
+// UserClient represents a per-user MCP client with multiple server connections
+type UserClient struct {
+	clients      map[string]*ServerConnection
+	toolDefs     map[string]ToolDefinition
+	lastActivity time.Time
+	userID       string
+	log          pluginapi.LogService
+}
+
+// ConnectToAllServers initializes connections to all provided servers
+func (c *UserClient) ConnectToAllServers(servers map[string]ServerConfig) error {
+	// Initialize clients for each server
+	for serverID, serverConfig := range servers {
+		if serverConfig.BaseURL == "" {
+			c.log.Warn("Skipping MCP server with empty BaseURL", "serverID", serverID)
+			continue
+		}
+
+		if err := c.connectToServer(context.Background(), serverID, serverConfig); err != nil {
+			c.log.Error("Failed to connect to MCP server", "userID", c.userID, "serverID", serverID, "error", err)
+			continue
+		}
+	}
+
+	// If no servers were successfully connected, return error
+	if len(c.clients) == 0 {
+		c.log.Warn("No MCP servers were successfully connected for user", "userID", c.userID)
+		return fmt.Errorf("no MCP servers were successfully connected")
+	}
+
+	// Update last activity time
+	c.lastActivity = time.Now()
+
+	return nil
+}
+
+// connectToServer establishes a connection to a single server and registers its tools
+func (c *UserClient) connectToServer(ctx context.Context, serverID string, serverConfig ServerConfig) error {
+	var opts []client.ClientOption
+	headers := make(map[string]string)
+	headers[MMUserIDHeader] = c.userID
+	if serverConfig.Headers != nil {
+		maps.Copy(headers, serverConfig.Headers)
+	}
+	opts = append(opts, client.WithHeaders(serverConfig.Headers))
+
+	sseClient, err := client.NewSSEMCPClient(serverConfig.BaseURL, opts...)
+	if err != nil {
+		return fmt.Errorf("failed to create MCP client: %w", err)
+	}
+
+	// Ensure client is closed on error
+	success := false
+	defer func() {
+		if !success {
+			sseClient.Close()
+		}
+	}()
+
+	if startErr := sseClient.Start(ctx); startErr != nil {
+		return fmt.Errorf("failed to start MCP client: %w", startErr)
+	}
+
+	initResult, err := sseClient.Initialize(ctx, mcp.InitializeRequest{})
+	if err != nil {
+		return fmt.Errorf("failed to initialize MCP client: %w", err)
+	}
+
+	c.log.Debug("MCP client initialized successfully",
+		"userID", c.userID,
+		"serverID", serverID,
+		"serverInfo", initResult.ServerInfo)
+
+	serverClient := &ServerConnection{
+		client:   sseClient,
+		serverID: serverID,
+		tools:    make(map[string]mcp.Tool),
+	}
+	c.clients[serverID] = serverClient
+
+	// List and register available tools
+	result, err := sseClient.ListTools(ctx, mcp.ListToolsRequest{})
+	if err != nil {
+		return fmt.Errorf("failed to list tools: %w", err)
+	}
+
+	// Store the tools for this server
+	for _, tool := range result.Tools {
+		serverClient.tools[tool.Name] = tool
+
+		// Check for tool name conflicts across servers
+		if existingTool, exists := c.toolDefs[tool.Name]; exists {
```

Comment:
I'm not 100% about this. maybe we should be able to have not-conflicting version of this, something like adding a the serverID as prefix to the tool if there is a conflict. That way, you can use different tools (based on the description) even if they conflict in the name.

Reply:
We probably should do something here. The issue is the tool names are the prompt engineering of the tools they are kinda important so we don't want to mess them up unless we have to. 

Not going to resolve this in this initial PR, but I will create a ticket for this.